pkg/neonvm/ipam/ipam.go

```diff
@@ -42,32 +37,85 @@ const (
 	IpamRequestTimeout = 10 * time.Second
 
 	// DatastoreRetries defines how many retries are attempted when reading/updating the IP Pool
-	DatastoreRetries             = 5
-	DatastoreRetriesDelay        = 100 * time.Millisecond
-	DefaultLeaderLeaseDurationMs = 3000
-	DefaultLeaderRenewDeadlineMs = 2500
-	DefaultLeaderRetryPeriodMs   = 2000
+	DatastoreRetries      = 5
+	DatastoreRetriesDelay = 100 * time.Millisecond
+
+	// CleanupInterval defines how often the IPAM cleaner is run
+	CleanupInterval = 15 * time.Minute
 )
 
+var ErrAgain = errors.New("IPAM concurrency limit reached. Try again later.")
+
 type Temporary interface {
 	Temporary() bool
 }
 
 type IPAM struct {
 	Client
 	Config IPAMConfig
+
+	mu                 sync.Mutex
+	concurrencyLimiter *semaphore.Weighted
+}
+
+func (i *IPAM) AcquireIP(ctx context.Context, vmName types.NamespacedName) (net.IPNet, error) {
+	ip, err := i.runIPAM(ctx, getAcquireAction(vmName))
+	if err != nil {
+		return net.IPNet{}, fmt.Errorf("failed to acquire IP: %w", err)
+	}
+	return ip, nil
+}
+
+func (i *IPAM) ReleaseIP(ctx context.Context, vmName types.NamespacedName) (net.IPNet, error) {
+	ip, err := i.runIPAM(ctx, getReleaseAction(vmName))
+	if err != nil {
+		return net.IPNet{}, fmt.Errorf("failed to release IP: %w", err)
+	}
+	return ip, nil
+}
+
+func (i *IPAM) runCleanup(ctx context.Context, namespace string) error {
```

Comment:
question: Do we actually need the cleanup action? If it's to handle previously leaked IP addresses from earlier versions of the code, could we just disable overlay IPs and manually reset the IP pool object? If we're worried about rare bugs in the future, we could add an HTTP endpoint to manually run `ReleaseIP(...)` for a particular VM.

Reply:
Fair enough, replaced it with [`04be1a6` (#1278)](https://github.com/neondatabase/autoscaling/pull/1278/commits/04be1a67a66a72b6d21bca254e41f62bee8d4982)